### ISSUE TITLE
[RyuJIT] Improve VZEROUPPER insertion

### ISSUE THREAD

**Issue by fiigii** [dotnet/coreclr#21062]
Currently, RyuJIT inserts VZEROUPPER instruction very conservatively https://github.com/dotnet/coreclr/issues/7761#issuecomment-272287724, which inserts VZEROUPPER at function prologue and epilogue when the method uses 128-bit or 256-bit AVX instructions. Recently, we come across certain scenarios having CQ problems from VZEROUPPER, e.g., https://github.com/dotnet/coreclr/issues/20820, https://github.com/dotnet/coreclr/issues/21055#issuecomment-439465776, etc.

The better insertion strategy that is adopted most of the native compilers is:
1. Adding VZEROUPPER instruction after **256-bit AVX instructions** are executed (only epilogue, ).  exceptions are if any arguments supplied or return values are in YMM/ZMM registers with `__vectorcall` that we may support in the future).
2. Adding VZEROUPPER before any function call that might execute legacy SSE code (P/Invoke, calling VM, and AOTed code).

**Comment by fiigii** [comment]
@CarolEidt @jkotas @AndyAyersMS @tannergooding

**Comment by CarolEidt** [comment]
To clarify, this would mean that in the cases where we are currently inserting a vzeroupper in both prolog and epilog, we would change it to epilog only, and then before certain function calls. The problem with the latter is that I'm not sure we have reliable information about whether code is precompiled (we do know about the P/Invoke case). As we continue down the tiered jitting path, we could have cases where the called method changes from AOT to JIT'd.

Also, as a shorter-term adjustment, would it make sense to stop emitting the vzeroupper in the prolog?

**Comment by fiigii** [comment]
> To clarify, this would mean that in the cases where we are currently inserting a vzeroupper in both prolog and epilog, we would change it to epilog only

For methods that use 256-bit AVX instructions, yes. For methods that use 128-bit AVX instructions **only**, we do not need any vzeroupper neither in prolog nor epilog. 

>  The problem with the latter is that I'm not sure we have reliable information about whether code is precompiled (we do know about the P/Invoke case). 

Agree. Actually, RyuJIT does not generate vzeroupper for call-sites to precompiled code, and it just handles P/Invoke call-sites now.

> Also, as a shorter-term adjustment, would it make sense to stop emitting the vzeroupper in the prolog?

Yes, I think so.

**Comment by fiigii** [comment]
> As we continue down the tiered jitting path

So, in the future, all the precompiled code will be JITed at Tier1, right?

**Comment by fiigii** [comment]
>> Also, as a shorter-term adjustment, would it make sense to stop emitting the vzeroupper in the prolog?

>Yes, I think so.

To clarify, once we can corrently hanlde all AVX->SSE call-sites (i.e., inserting VZEROUPPER before any function call that might execute legacy SSE code), methods that use 256-bit AVX instructions would not need VZEROUPPER in prolog. Because pure SSE code is in "clean upper state", and "clean upper state"-> AVX code does not have any penalty in user code.

**Comment by CarolEidt** [comment]
@fiigii - can you explain the context in which we *would* want to add a VZEROUPPER in the prolog currently?

**Comment by fiigii** [comment]
I do not know the exact reason that RyuJIT inserts VZEROUPPER in the prolog currently, that code was added before I joined CoreCLR 😄 . But I guess it is because we are not able to insert VZEROUPPER before calling to SSE code (i.e., JIT code -> crossge code) sometimes, which makes 256-bit AVX code not sure its caller is "clean upper state" or not.

Assume there are three methods, _A_ and _B_ have 256-bit AVX code and compiled by JIT, _C_ has FP code and compiled by AOT. _A_ calls to _C_ and _C_ calls to _B_.

The correct way should be: _A_ -> VZEROUPPER(before call-site) -> _C_(clean upper state) -> _B_ -> VZEROUPPER(B's epilogue) -> _C_(clean upper state) -> _A_.

But if we cannot insert the first VZEROUPPER, it would become: _A_ -> _C_(dirty upper state, penalty) -> (penalty) _B_ -> VZEROUPPER(B's epilogue) -> _C_ -> _A_.

Then inserting VZEROUPPER in _B_'s prologue could make it better: _A_ -> _C_(dirty upper state, penalty) -> VZEROUPPER(B's prologue) _B_ -> VZEROUPPER(B's epilogue) -> _C_(clean upper state) -> _A_.

**Comment by CarolEidt** [comment]
@fiigii - thanks. What I don't understand is this bit: C(dirty upper state, penalty) -> (penalty) B

Is it the case that the AVX code gets a penalty when transitioning from SSE with dirty upper state? I guess that makes sense, but it wasn't clear to me previously (or it was, but I have since forgotten it!)

**Comment by fiigii** [comment]
> Is it the case that the AVX code gets a penalty when transitioning from SSE with dirty upper state? 

Yes, SSE -> AVX transition penalty only happens when the SSE code is in dirty upper state, which dirty upper is from previous AVX->SSE transition. And AVX->SSE transition always happens if we do not insert VZEROUPPER before 256-bit-AVX -> SSE.

**Comment by CarolEidt** [comment]
Great - thanks!

**Comment by tannergooding** [comment]
It's probably handy to leave these diagrams here, which are from the architecture optimization manual:

Skylake and newer micro-architectures:
![image](https://user-images.githubusercontent.com/1263030/48019732-f73ffb00-e0e8-11e8-9159-910c78991927.png)

Older micro-architectures:
![image](https://user-images.githubusercontent.com/1263030/48019799-248ca900-e0e9-11e8-986f-6e0deefec298.png)

**Comment by tannergooding** [comment]
@fiigii and I had a brief discussion on this here: https://github.com/dotnet/coreclr/pull/20788#discussion_r230593796

**Comment by CarolEidt** [comment]
Now if I can just remember to come back here when I have these questions ...

**Comment by fiigii** [comment]
> and I had a brief discussion on this here: #20788 (comment)

In that discussion, I worried about 128-bit AVX <-> SSE penalty (`Preserved Non-INIT Upper state` in the older micro-architecture diagram). But I consulted our native compiler experts, they said that state should be correctly handled by OS code, so we do not need any VZEROUPPER in 128-bit only AVX code.

**Comment by CarolEidt** [comment]
I would be inclined to put this in "Future" but also flag it as "Up for Grabs" as I don't think the implementation is terribly complex, or impactful.
